### PR TITLE
Adding Moe (Mixture of Experts) layer 

### PR DESCRIPTION
**Pull request by LiuHaolan**
- 混合专家模型(Mixture of Experts) 的实现
 - 在cifar10数据集上测试精度和pytorch开源实现对齐

### CODE REVIEW

File: MoE/moe.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        zeros = torch.zeros(self._gates.size(
+            0), expert_out[-1].size(1), requires_grad=True)
+
+        # spanning a index matrix
+        batch_index = np.zeros([stitched.shape[0], stitched.shape[1]])
+
+        for i in range(stitched.shape[0]):
+            batch_index[i, :] = np.ones(
+                batch_index.shape[1])*self._batch_index[i].item()
+        batch_index_ = flow.Tensor(batch_index)
```
Comment (hjchen2) [inline MoE/moe.py:136]:
```
batch_index_ = flow.Tensor(batch_index，device=stitched.device)
```

File: MoE/moe.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+from mlp import MLP
+import numpy as np
+
+from lenet import LeNet
+
+"""
+import math
+import oneflow
+import oneflow as flow
+import oneflow as torch
```
Comment (hjchen2) [inline MoE/moe.py:23]: 这两个只保留一个吧，最好保留```import oneflow as flow```

File: MoE/moe.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        Returns:
+          a `Tensor` with shape `[batch_size, <extra_output_dims>]`.
+        """
+        # apply exp to expert outputs, so we are not longer in log space
+
+        stitched = torch.cat(expert_out, 0).exp()
+
+        if multiply_by_gates:
+            stitched = stitched.mul(self._nonzero_gates)
+        zeros = torch.zeros(self._gates.size(
```
Comment (hjchen2) [inline MoE/moe.py:127]: source op都可以指定device的，比如
```
zeros = torch.zeros(self._gates.size(0), expert_out[-1].size(1), requires_grad=True, device=stitched.device)
```

**Comment by hjchen2** [comment]
fast moe开源的代码里支持组建式搭建moe模型，可以将一个普通的mlp转换成moe，可以借鉴一下。
```python
model = ...

from fmoe.megatron import fmoefy
model = fmoefy(model, num_experts=<number of experts per worker>)

train(model, ...)
```

https://github.com/laekov/fastmoe

**Comment by Ldpe2G** [comment]
需要在 reamd model list 里面加一下，应该是个新的类别
https://github.com/Oneflow-Inc/models/blob/main/README.md#model-list
https://github.com/Oneflow-Inc/models/blob/main/README_zh-CN.md#%E6%A8%A1%E5%9E%8B%E7%9B%AE%E5%BD%95

File: NLP/MoE/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+model = ...
+
+from fmoe.megatron import fmoefy
+model = fmoefy(model, num_experts)
+
+train(model, ...)
+
+
+```
```
Comment (Ldpe2G) [inline NLP/MoE/README.md:24]: 上面这段代码在 `cifar10_example.py` 没找到，这里换成怎么跑 cifar10_example.py 脚本吧，还要再加一个 train.sh 的脚本，用户直接运行 train.sh 就行了，比如数据集下载的路径配置，模型保存的路径啥的
可参考：https://github.com/Oneflow-Inc/models/blob/main/Vision/classification/image/mobilenetv2/train.sh

File: NLP/MoE/cifar10_example.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+import oneflow as torch
+import oneflow as flow
+import oneflow.nn as nn
+import oneflow.nn.functional as F
+import oneflow.optim as optim
+from mlp import MLP
+from moe import MoE
+
+import flowvision as vision
```
Comment (Ldpe2G) [inline NLP/MoE/cifar10_example.py:11]: 依赖了除  oneflow 之外的库， 需要再加一个 requirements.txt 文件

File: NLP/MoE/cifar10_example.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+transform = vision.transforms.Compose(
+    [vision.transforms.ToTensor(),
+     vision.transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+trainset = vision.datasets.CIFAR10(root='./data', train=True,
```
Comment (Ldpe2G) [inline NLP/MoE/cifar10_example.py:17]: 路径啥的这些可变的参数都通过 命令行传入吧

File: NLP/MoE/cifar10_example.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                                       download=True, transform=transform)
+testloader = flow.utils.data.DataLoader(testset, batch_size=64,
+                                        shuffle=False, num_workers=1)
+
+classes = ('plane', 'car', 'bird', 'cat',
+           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')
+
+
+device = flow.device('cuda')
+expert_network = MLP(input_size=3072, output_size=10, hidden_size=256)
```
Comment (Ldpe2G) [inline NLP/MoE/cifar10_example.py:32]: input_size 这个也是可配置的？

Comment (LiuHaolan) [reply NLP/MoE/cifar10_example.py:32]: 是可配置的，但是取决于用户选用当作expert network的input_size，感觉当作命令行参数不方便。我之后看看能不能选用一个更简洁的接口。

**Review by Ldpe2G — APPROVED** [review]
